### PR TITLE
feat: Support controller horizontal scaling

### LINKED ISSUE #4284: Controller horizontal scaling
# Summary

The application controller is the only Argo CD component that cannot have multiple replicas. Allow to scale up Application Controller to increase the number of clusters on Argo CD can manage. 

# Motivation

Currently number of managed clusters is limited by the amount of memory allowed to one controller instance. It is impossible to grow memory infinitely. Instead, it should be possible to shared clusters among multiple controller replicas.

# Proposal

The Application controller should be converted to a StatefulSet and each replica should handle only a subset of clusters and related applications.

**Sharding method**

Cluster size is not known ahead of time so it is not easy to implement a good sharding method that guarantees well-balanced distribution. Looks like we have two possible options:

1. Dynamic sharding based on cluster size. Controllers should "claim" clusters one by one, evaluate the number of resources in each cluster. This approach should provide better clusters distribution but significantly more difficult to implement and might require some coordination between controller replicas.

2. Simple sharding.  Sharding function might looks like `shard = hash(clusterSecret.metadata.UID) % replicas`.  This should work well if the controller manages a lot of mid-size clusters but won't work if there are several very large clusters, different from others. To address that it should be possible to manually "assign" cluster to the shard.

### PR DESCRIPTION
**Pull request by alexmt**
Closes https://github.com/argoproj/argo-cd/issues/4284

* [x] Implementation
* [x] Tests
* [x] Docs

### CODE REVIEW

File: controller/sharding/sharding.go [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+	if err != nil {
+		return 0, err
+	}
+	parts := strings.Split(hostname, "-")
+	if len(parts) == 0 {
+		return 0, fmt.Errorf("hostname should ends with shard number separated by '-' but got: %s", hostname)
+	}
+	shard, err := strconv.Atoi(parts[len(parts)-1])
+	if err != nil {
+		return 0, fmt.Errorf("hostname should ends with shard number separated by '-' but got: %s", hostname)
```
Comment (jsoref) [inline controller/sharding/sharding.go:24]:
```suggestion
		return 0, fmt.Errorf("hostname should end with shard number separated by '-' but got: %s", hostname)
	}
	shard, err := strconv.Atoi(parts[len(parts)-1])
	if err != nil {
		return 0, fmt.Errorf("hostname should end with shard number separated by '-' but got: %s", hostname)
```

**Comment by yydzhou** [comment]
@alexmt will the " horizontal scaling" feature be available in next release 1.18？ we have more than 700 applications now and the number keeps growing very fast. scalability of the controller is highly desired.

**Comment by alexmt** [comment]
Hello @yydzhou . Yes, it will be available in 1.8 release.

**Comment by alexmt** [comment]
Looks like it is ready to go. @jessesuen , @jannfis  can you take a look ?

**Review by jessesuen — APPROVED** [review]

File: cmd/argocd/commands/cluster.go [hunk 2]
```diff
@@ -137,6 +141,7 @@ func NewClusterAddCommand(clientOpts *argocdclient.ClientOptions, pathOpts *clie
 	command.Flags().StringVar(&systemNamespace, "system-namespace", common.DefaultSystemNamespace, "Use different system namespace")
 	command.Flags().StringArrayVar(&namespaces, "namespace", nil, "List of namespaces which are allowed to manage")
 	command.Flags().StringVar(&name, "name", "", "Overwrite the cluster name")
+	command.Flags().Int64Var(&shard, "shard", -1, "Optional cluster shard number")
```
Comment (jsoref) [inline cmd/argocd/commands/cluster.go:144]: I'm not sure that `optional` adds value here. none of the other flags use it.

File: manifests/ha/base/overlays/argocd-application-controller-statefulset.yaml [hunk 3]
```diff
@@ -14,4 +14,4 @@ spec:
         - --operation-processors
         - "10"
         - --redis
-        - "argocd-redis-ha-haproxy:6379"
+        - "argocd-redis-ha-haproxy:6379"
```
Comment (jsoref) [inline manifests/ha/base/overlays/argocd-application-controller-statefulset.yaml:17]: Can you add the missing new-line at eof?

File: controller/sharding/sharding.go (hunk 1 above)
Comment (jsoref) [reply controller/sharding/sharding.go:24]: This is marked as resolved but doesn't appear to be?

File: cmd/argocd/commands/cluster.go (hunk 2 above)
Comment (alexmt) [reply cmd/argocd/commands/cluster.go:144]: Agree. Removed optional and added more details.

File: manifests/ha/base/overlays/argocd-application-controller-statefulset.yaml (hunk 3 above)
Comment (alexmt) [reply manifests/ha/base/overlays/argocd-application-controller-statefulset.yaml:17]: Done.

File: controller/sharding/sharding.go [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+	}
+	return shard, nil
+}
+
+func getShardByID(id string, replicas int) int {
```
[the remark this answers is not in the document]
Comment (alexmt) [reply controller/sharding/sharding.go:29]: Done.

**Comment by jannfis** [comment]
Just a heads-up, this PR will break self-updating Argo CD due to moving application-controller from deployment to statefulset (deployment will be pruned, and application-controller is gone then).

**Comment by edsonmarquezani** [comment]
Very interested in this new feature, but I already have one Argo per cluster. I would like to shard inside the cluster based on applications, not clusters. Is it possible?

**Comment by edsonmarquezani** [comment]
@alexmt Would you mind to tell me if this also applies for single cluster ArgoCD configurations? I mean, will each controller sync 1/N applications (where N is the number of replicas)?

**Comment by edsonmarquezani** [comment]
I tested it myself. The answer is: NO. Sharding strategy is by cluster only, which is a shame. =/ I have a cluster big enough to need to scale the controller.

**Pull request merged**